### PR TITLE
Split tests into separate functions for parallelization

### PR DESCRIPTION
**Pull request by agoncharov-reef**
Now Pawel will look at this and say "why so many fixtures?"

### CODE REVIEW

File: noxfile.py [hunk 1]
```diff
@@ -153,7 +153,7 @@ def integration(session):
     """Run integration tests."""
     install_myself(session)
     session.install(*REQUIREMENTS_TEST)
-    session.run('pytest', '-s', *session.posargs, 'test/integration')
+    session.run('pytest', '-s', '-n', 'auto', *session.posargs, 'test/integration')
```
Comment (ppolewicz) [inline noxfile.py:156]: this is appropriate for CPU-bound unit tests, but not for io-bound integration tests. For integration tests it should be much higher, think `5 * multiprocessing.cpu_count()`

File: test/integration/conftest.py [hunk 2]
```diff
@@ -115,10 +122,11 @@ def b2_tool(application_key_id, application_key, realm, this_run_bucket_name_pre
...
         this_run_bucket_name_prefix,
     )
-    tool.reauthorize(check=True)
+    tool.reauthorize(check=True)  # reauthorize for the first time (with check)
     return tool
 
 
 @pytest.fixture(scope='function', autouse=True)
 def auto_reauthorize(request, b2_tool):
+    """ Automatically reauthorize for each test (without check) """
```
Comment (ppolewicz) [inline test/integration/conftest.py:126]: do we really need to do it for each test? takes time

**Review by ppolewicz — COMMENTED** [review]
first commit does:
- renaming of test_bucketinfo to get_bucketinfo
- renaming of test functions
- moving code around without modification
- fixing comment lines
- reformatting code

because I don't know which commit is just moving code and which is modifying it, I have to audit the whole thing which takes a ton of time and causes reporting of issues that you did not introduce in the code. If you'd split it up to micro commits with single type change annotated properly, the review would be much easier.

Second commit fixes formatting from the first commit. Should just be `git commit -a --amend` if it's the next commit after making the change that needs to be fixed, because it doesn't mess up any history and then I wouldn't review bad stuff in the first commit.

File: test/integration/conftest.py [hunk 3]
```diff
@@ -27,5 +34,88 @@ def sut(request):
...
+
+
+@pytest.fixture(scope='session')
+def realm() -> str:
+    yield environ.get('B2_TEST_ENVIRONMENT', 'production')
+
+
+@pytest.fixture(scope='function')
+def bucket_name(b2_api):
+    yield b2_api.create_bucket()
```
[the remark this answers is not in the document]
Comment (agoncharov-reef) [reply test/integration/conftest.py:51]: No, it's fine: https://github.com/reef-technologies/B2_Command_Line_Tool/blob/improve-integration-tests/test/integration/test_b2_command_line.py#L267

File: test/integration/conftest.py [hunk 4]
```diff
@@ -27,5 +34,88 @@ def sut(request):
...
+    yield Api(
+        application_key_id, application_key, realm, general_bucket_name_prefix, this_run_bucket_name_prefix
+    )
+
+
+@pytest.fixture(scope='module', autouse=True)
+def auto_clean_buckets(b2_api):
+    b2_api.clean_buckets()
+    yield
+    b2_api.clean_buckets()
```
[the remark this answers is not in the document]
Comment (agoncharov-reef) [reply test/integration/conftest.py]: It uses to be called before and after each test: https://github.com/reef-technologies/B2_Command_Line_Tool/blob/improve-integration-tests/test/integration/test_b2_command_line.py#L2650
Now it's called before the entire testing, and afterwards. In fact, we could delete the "pre-cleanup", wdyt?

File: test/integration/conftest.py [hunk 5]
```diff
@@ -27,5 +34,88 @@ def sut(request):
...
+        realm,
+        this_run_bucket_name_prefix,
+    )
+    tool.reauthorize(check=True)
+    return tool
+
+
+@pytest.fixture(scope='function', autouse=True)
+def auto_reauthorize(request, b2_tool):
+    b2_tool.reauthorize(check=False)
```
[the remark this answers is not in the document]
Comment (agoncharov-reef) [reply test/integration/conftest.py:127]: Yes we do, but I haven't found a way to port this line, when we `reauthorize(check=True)` for first test, and `reauthorize(check=False)` for all subsequent tests: https://github.com/reef-technologies/B2_Command_Line_Tool/blob/improve-integration-tests/test/integration/test_b2_command_line.py#L2658

File: test/integration/helpers.py [hunk 6]
```diff
@@ -0,0 +1,457 @@
...
+            if BUCKET_CREATED_AT_MILLIS in bucket.bucket_info:
+                delete_older_than = current_time_millis() - ONE_HOUR_MILLIS
+                this_bucket_creation_time = bucket.bucket_info[BUCKET_CREATED_AT_MILLIS]
+                if int(this_bucket_creation_time) < delete_older_than:
+                    return True, f"{this_bucket_creation_time} < {delete_older_than}"
+            else:
+                return True, 'undefined ' + BUCKET_CREATED_AT_MILLIS
+        return False, ''
+
+    def clean_buckets(self):
```
[the remark this answers is not in the document]
Comment (agoncharov-reef) [reply test/integration/helpers.py:109]: Yep, didn't touch it at all, just moved here

File: test/integration/conftest.py (hunk 2 above)
Comment (agoncharov-reef) [reply test/integration/conftest.py:126]: Copied for old setup. I have no clue why it's called before each test: https://github.com/reef-technologies/B2_Command_Line_Tool/blob/improve-integration-tests/test/integration/test_b2_command_line.py#L2658

File: test/integration/conftest.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ppolewicz) [reply test/integration/conftest.py:51]: I think it should return a Bucket object and in this function you should return the `.name` field. `b2_api.create_bucket()` returning bucket name is super confusing.

File: test/integration/test_b2_command_line.py [hunk 7]
```diff
@@ -443,15 +442,12 @@ def sync_up_helper(b2_tool, bucket_name, dir_, encryption=None):
...
-
         file_versions = b2_tool.list_file_versions(bucket_name)
         should_equal([], file_version_summary(file_versions))
 
-        write_file(p('a'), b'hello')
-        write_file(p('b'), b'hello')
-        write_file(p('c'), b'hello')
+        write_file(os.path.join(dir_path, 'a'), b'hello')
+        write_file(os.path.join(dir_path, 'b'), b'hello')
+        write_file(os.path.join(dir_path, 'c'), b'hello')
```
Comment (ppolewicz) [inline test/integration/test_b2_command_line.py:450]: noooo this is more ugly now! It would only be better if it used pathlib with the `/` operator

**Comment by ppolewicz** [comment]
Some things could be better, but lets not fix everything in one PR, this one is big enough. Ready to submit to upstream.

**Pull request closed without merging**